presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java

```diff
@@ -106,153 +130,178 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
-                    return new LdapPrincipal(user);
-                }
-            }, servletResponse);
+            return new LdapPrincipal(user);
         }
-        catch (AuthenticationException e) {
-            // log the failure since the exception goes directly to the client
+        catch (javax.naming.AuthenticationException e) {
             log.debug("Authentication failed for user %s. Invalid credentials: %s", user, e.getMessage());
-            throw new RuntimeException("Invalid credentials: " + e.getMessage());
+            throw new AuthenticationException(UNAUTHORIZED, "Invalid credentials", e);
+        }
+        catch (NamingException e) {
+            throw new AuthenticationException(INTERNAL_SERVER_ERROR, "Authentication failed", e);
         }
         finally {
-            try {
-                if (context != null) {
-                    context.close();
-                }
-            }
-            catch (NamingException ignored) {
-            }
+            closeContext(context);
         }
     }
 
-    private void checkForGroupMembership(String searchFilterPattern, String user, DirContext context)
+    private Map<String, String> createEnvironment(Credentials credentials)
     {
+        ImmutableMap.Builder<String, String> environment = ImmutableMap.builder();
+        environment.putAll(basicEnvironment);
+        environment.put(SECURITY_AUTHENTICATION, "simple");
+        environment.put(SECURITY_PRINCIPAL, createPrincipal(credentials.getUser()));
+        environment.put(SECURITY_CREDENTIALS, credentials.getPassword());
+        return environment.build();
+    }
+
+    private String createPrincipal(String user)
+    {
+        return genericLdapBinder.getUserBindSearchPattern().replaceAll("\\$\\{USER\\}", user);
+    }
+
+    private static InitialDirContext createDirContext(Map<String, String> environment)
+            throws NamingException
+    {
+        return new InitialDirContext(new Hashtable<>(environment));
+    }
+
+    private void checkForGroupMembership(String user, DirContext context)
+            throws AuthenticationException
+    {
+        Optional<String> pattern = genericLdapBinder.getGroupAuthorizationSearchPattern();
+        if (!pattern.isPresent()) {
+            return;
+        }
+
         checkState(serverConfig.getUserBaseDistinguishedName() != null, "Base distinguished name (DN) for user %s is null", user);
 
-        try {
-            String searchFilter = searchFilterPattern.replaceAll("\\$\\{USER\\}", user);
-            GroupAuthorizationFilter groupFilter = new GroupAuthorizationFilter(serverConfig.getUserBaseDistinguishedName(), searchFilter);
+        String searchFilter = pattern.get().replaceAll("\\$\\{USER\\}", user);
+        String userBaseDistinguishedName = serverConfig.getUserBaseDistinguishedName();
+        String groupName = firstNonNull(serverConfig.getLdapServerType().equals(GENERIC) ? serverConfig.getGroupDistinguishedName() : "", "");
```

Comment:
This logic is a) pretty confusing and b) different than what was used before:
`String groupNameIfPresent = firstNonNull(serverConfig.getLdapServerType().equals(GENERIC) ? null : serverConfig.getGroupDistinguishedName(), "");`
What is it trying to do? It may be worthwhile to extract a method here and actually write out the if statements rather than using a ternary + firstNonNull